packages/medusa/src/services/order.js

```diff
@@ -0,0 +1,447 @@
+      )
+    }
+
+    const { shipping_methods, items } = order
+
+    // if only one shipping method is attached to the order, we simply create
+    // the order for the single provider given in the method
+    if (shipping_methods.length === 1) {
+      const fulfillmentProviderId = shipping_methods[0].provider_id
+      const fulfillmentProvider = this.fulfillmentProviderService_.retrieveProvider(
+        fulfillmentProviderId
+      )
+
+      shipping_methods[0].items = items
+
+      await fulfillmentProvider.createOrder(shipping_methods[0].data, items)
+
+      return this.orderModel_.updateOne(
+        {
+          _id: orderId,
+        },
+        {
+          $set: { fulfillment_status: "fulfilled" },
+        }
+      )
+    }
+
+    // partition order items to their dedicated shipping method
+    order.shipping_methods = await this.partitionItems_(shipping_methods, items)
+
+    await Promise.all(
+      order.shipping_methods.map(method => {
+        const provider = this.fulfillmentProviderService_.retrieveProvider(
+          method.provider_id
+        )
+        provider.createOrder(method.data, method.items)
+      })
+    )
+
+    return this.orderModel_.updateOne(
+      {
+        _id: orderId,
+      },
+      {
+        $set: { fulfillment_status: "fulfilled" },
+      }
+    )
+  }
+
+  /**
+   * Return either the entire or part of an order.
+   * @param {string} orderId - the order to return.
+   * @param {string[]} lineItems - the line items to return
+   * @return {Promise} the result of the update operation
+   */
+  async return(orderId, lineItems) {
+    const order = await this.retrieve(orderId)
+
+    if (
+      order.fulfillment_status === "not_fulfilled" ||
+      order.fulfillment_status === "returned"
+    ) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_ALLOWED,
+        "Can't return an unfulfilled or already returned order"
+      )
+    }
+
+    if (order.payment_status !== "captured") {
+      throw new MedusaError(
+        MedusaError.Types.NOT_ALLOWED,
+        "Can't return an order with payment unprocessed"
+      )
+    }
+
+    const { provider_id, data } = order.payment_method
+    const paymentProvider = this.paymentProviderService_.retrieveProvider(
+      provider_id
+    )
+
+    const amount = this.totalsService_.getRefundTotal(order, lineItems)
+    await paymentProvider.refundPayment(data, amount)
+
+    lineItems.map(item => {
+      const returnedItem = order.items.find(({ _id }) => _id === item._id)
+      if (returnedItem) {
+        returnedItem.returned_quantity = item.quantity
+      }
+    })
+
+    const fullReturn = order.items.every(
+      item => item.quantity === item.returned_quantity
+    )
+
+    return this.orderModel_.updateOne(
+      {
+        _id: orderId,
+      },
+      {
+        $set: {
+          items: order.items,
+          fulfillment_status: fullReturn ? "returned" : "partially_fulfilled",
+        },
+      }
+    )
+  }
+
```

Comment:
I think we should include an `archive` method which just checks if the order has been captured and fulfilled, i.e. the status of the order is `completed` or `refunded`. If that is the case we are allowed to mark the order as `archived`.

Reply:
Indeed! Good point